File: rotatedArraySearch.java
   Time Complexity-O(log n)
   Space Complexity-O(1)
   public int search(int[] nums, int target) {
        int n=nums.length;
        int l=0,h=n-1;
        while(l<h){
            int m=(l+h)/2;
            if(nums[m]>nums[h]) l=m+1;
            else h=m;
        }
      
        int rot=l;
        l=0;h=n-1;
        while(l<=h){
            int m=(l+h)/2;
            int adjustedMid=(m+rot)%n;
            if(nums[adjustedMid]==target)return adjustedMid;
            if(nums[adjustedMid]<target)l=m+1;
            else h=m-1;
        }
        return -1;
    }

File: rotatedArraySearch.cpp
// Time Complexity :O(logN)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :

#include <iostream>
#include <vector>

using namespace std;
class Solution {
public:
    int search(vector<int>& nums, int target) {
        if(nums.empty() || nums.size() == 0){
            return -1;
        }
        int low = 0;
        int high = nums.size() - 1;
        while(low <= high){
            int mid = low + (high-low) / 2;
            if(target == nums[mid]){
                return mid;
            }
            else if(nums[low] <= nums[mid]){ //left portion
                if(target >= nums[low] && target < nums[mid]){
                    high = mid - 1;
                }
                else{
                    low = mid + 1;
                }
            }
            else{
                if(target > nums[mid] && target <= nums[high]){
                    low = mid + 1;
                }
                else{
                    high = mid - 1;
                }
            }
        }
        return -1;
    }
};

int main(){
    class Solution s;
    vector<int> matrix = {4,5,6,0,1,2,3};
    int value = 1;
    int x = s.search(matrix, value);
    std::cout << x << std::endl;
    return 0;
}
